### PR TITLE
Add setting to preserve cell metadata for copied cells

### PR DESCRIPTION
**Pull request by DenisaCG**
## References
This PR offers a solution for https://github.com/jupyterlab/jupyterlab/issues/13228.

## Code changes
The `preserveCellMetadataForCopiedCells` setting was added in `Notebook Settings` and default to false. When enabled, the copy, duplicate and cut workflows will keep the cell metadata as it is, instead of stripping the `deletable` field, as is the current behavior.

The `selectedCells` function was modified to account for the flag and keep all the cell metadata, only when enabled. Then all other relevant functions which call `selectedCells`, send `preserveMetadata` through as an argument.

As a note, the previous code was only removing the `deletable` cell metadata field,  as such the new setting only changes if this particular field is either removed or kept when the cell is selected and its subsequent actions. However, this setting can also be used for other metadata fields which might require the same behavior in the future.

Moreover, two tests were added for preserving the cell metadata, mirroring the existing ones which test the removal of the `deletable` field.

## User-facing changes
The user will have access to a new setting that will let them preserve the cell metadata for copied cells when enabled. Since the setting defaults to false, the users which do not have an interest in this behavior will not be affected or notice any difference in the default behavior.

The screencast below shows how the setting can be accessed and its impact on the metadata of the copied and pasted cells:

[Screencast from 05.06.2026 13:56:33.webm](https://github.com/user-attachments/assets/bd300004-6ea1-4f55-997c-897b95a91d41)

## Backwards-incompatible changes
None.

## AI usage

- **NO**: Some or all of the content of this PR was generated by AI.
- **YES**: The human author has carefully reviewed this PR and run this code (keep this PR "draft" until the answer is YES)
- AI tools and models used: None.

cc: @afshin

### CODE REVIEW

File: packages/notebook/test/actions.spec.ts [hunk 1]
```diff
@@ -1722,6 +1722,19 @@ describe('@jupyterlab/notebook', () => {
...
+      it('should preserve all cell metadata', () => {
+        const next = widget.widgets[1];
+        widget.select(next);
+        next.model.setMetadata('deletable', false);
+        NotebookActions.copy(widget, true);
+        const data = utils.clipboard.getData(JUPYTER_CELL_MIME) as JSONArray;
+        data.map(cell => {
+          expect(
+            ((cell as JSONObject).metadata as JSONObject).deletable
+          ).toBeFalsy();
```
Comment (MUFFANUJ) [inline packages/notebook/test/actions.spec.ts]: This assertion does not actually prove that `deletable` was preserved. `toBeFalsy()` also passes when the field is `undefined`, which is the current stripping behavior, and this selection includes the active cell whose `deletable` metadata was never set. Could we make this a strict regression test by either setting `deletable: false` on every copied cell or locating the modified cell, then asserting `expect(metadata).toHaveProperty('deletable', false)`? The same issue applies to the system clipboard assertion below.

**Review by MUFFANUJ — COMMENTED** [review]
This looks good in general few comments. 
Since the setting is advertised for copy, cut, and duplicate, could we add coverage for the non-copy paths as well? Right now, only `copy()` and `copyToSystemClipboard()` exercise `preserveMetadata`, so a regression in `cut()`, `cutToSystemClipboard()`, or this `duplicate()` path would not be caught. Tests for each path using `deletable: false` and a property assertion should do.

**Comment by DenisaCG** [comment]
Thank you for your review and suggestions, @MUFFANUJ!

I have updated the existing copy action tests to assert the `deletable` property as false directly and ensured all selected cells have their metadata field modified and checked accordingly. I have also included additional tests for the `cut` and `duplicate` actions: one to check that the `deletable` field is removed and another to check whether it preserves all metadata when the setting is activated. These set of tests should now cover any regression for all three actions.

File: packages/notebook/test/actions.spec.ts (hunk 1 above)
Comment (DenisaCG) [reply packages/notebook/test/actions.spec.ts]: Thank you for the suggestion, it has been addressed in https://github.com/jupyterlab/jupyterlab/pull/18994/commits/7aa6aabf5624f2102c8d808cfe72fb2d8ffa295a.

**Comment by MUFFANUJ** [comment]
This has picked up conflicts! @DenisaCG, can you resolve it?

File: packages/notebook/src/actions.tsx [hunk 2]
```diff
@@ -1398,48 +1398,60 @@ export namespace NotebookActions {
    * Copy the selected cell(s) data to a clipboard.
    *
    * @param notebook - The target notebook widget.
+   * @param preserveMetadata - Whether to preserve cell metadata.
    */
-  export function copy(notebook: Notebook): void {
-    Private.copyOrCut(notebook, false);
+  export function copy(
+    notebook: Notebook,
+    preserveMetadata: boolean = false
```
Comment (MUFFANUJ) [inline packages/notebook/src/actions.tsx:1405]: This is a non-blocking suggestion for adding more clarity: should we make this an options object instead of a positional boolean? Since `copy`, `cut`, and `duplicate` are public APIs, `NotebookActions.copy(widget, { preserveMetadata: true })` would be clearer and easier to extend than `NotebookActions.copy(widget, true)`.

File: packages/notebook-extension/schema/tracker.json [hunk 3]
```diff
@@ -898,6 +898,12 @@
       "type": "boolean",
       "default": false
     },
+    "preserveCellMetadataForCopiedCells": {
+      "title": "Preserve cell metadata of copied cells",
+      "description": "When enabled, the copied and pasted cells will preserve all their cell metadata.",
```
Comment (MUFFANUJ) [inline packages/notebook-extension/schema/tracker.json:903]: Could we narrow this description to mention `metadata.deletable`? In `Private.selectedCells()`, the copied cell already comes from `cell.model.toJSON()`, and the only metadata key conditionally changed is `deletable`, which is deleted when `preserveMetadata` is false. So “all their cell metadata” overstates what this setting changes.
